ui/src/components/FacetsGrid.js

```diff
@@ -1,26 +1,94 @@
-            cols={3}
-            cellHeight="auto"
-            padding={1}
-        >
-            {facetsList}
-        </GridList>
-    );
+class FacetsGrid extends Component {
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            facetsResponse: props.facetsResponse
+        };
+        this.searchFilters = new Map();
+        let apiClient = new ApiClient();
+        apiClient.basePath = '/api';
+        this.api = new FacetsApi(apiClient);
+        this.callback = function(error, data) {
+            if (error) {
+                console.error(error);
+                // TODO(alanhwang): Redirect to an error page
+            } else {
+                this.setState({facetsResponse: data});
+            }
+        }.bind(this);
+        this.updateFacets = this.updateFacets.bind(this);
+    }
+
+    render() {
+        const facetsList = this.state.facetsResponse.facets.map((facet) =>
+            <GridTile key={facet.name}>
+                <FacetCard
+                    facet={facet}
+                    count={this.state.facetsResponse.count}
+                    updateFacets={this.updateFacets}
+                    className='facetCard'/>
+            </GridTile>
+        );
+        return (
+            <GridList
+                className="gridList"
+                cols={3}
+                cellHeight="auto"
+                padding={1}
+            >
+                {facetsList}
+            </GridList>
+        );
+    }
+
+    /**
+     * Updates the selection for a single facet value and refreshes the facets data from the server.
+     * @param facetName string containing the name of the facet corresponding to this value
+     * @param facetValue string containing the name of this facet value
+     * @param isSelected bool indicating whether this facetValue should be added to or removed from the query
+     * */
+    updateFacets(facetName, facetValue, isSelected) {
+        if (this.searchFilters.get(facetName) === undefined && isSelected) {
+            this.searchFilters.set(facetName, [facetValue]);
+        } else if (this.searchFilters.get(facetName) !== undefined) {
+            if (isSelected && !this.searchFilters.get(facetName).includes(facetValue)) {
+                this.searchFilters.get(facetName).push(facetValue);
+            } else if (!isSelected) {
+                this.searchFilters.set(facetName, this.removeFacet(this.searchFilters.get(facetName), facetValue));
+            }
+        }
+        console.log(this.serialize(this.searchFilters));
+        this.api.facetsGet({filter: this.serialize(this.searchFilters)}, this.callback);
+    }
+
+    removeFacet(valueList, facetValue) {
+        for (let i = valueList.length-1; i >= 0; i--) {
+            if (valueList[i] === facetValue) {
+                valueList.splice(i, 1);
+                break;
+            }
+        }
+        return valueList;
+    }
+
+    serialize(searchFilters) {
```

Comment:
This is what I see:

filter=Age=30-39,60-69&Region=southeast

This is what we want:

filter=Age=30-39&filter=Age=60-69&filter=Region=southeast

See https://swagger.io/docs/specification/serialization/#query

On second thought, probably this is better (up to you):

filter=Age=30-39,Age=60-69,Region=southeast

You would need to update api.yaml.

(Added this to things we should unit test in #7)

Reply:
Actually, we should be using Swagger-generated Facet, FacetValue, etc. Then we don't have to worry about serialization.

Reply:
Current implementation is your second suggestion.

Your idea to use the swagger-generated models is interesting, but I'm not sure how to do this practically. It would be nice if those came with handy serializing/deserializing utilities like protos, but currently when I encode something like a Gender facet with one value female I get `Gender&filter[values]=female`. As far as I can tell, it's not any kind of standard format that we can decode in Python, so I'm not sure how to go about deserializing this on the server side.

Reply:
Oops, never mind about the swagger. I forgot Facet/FacetValue isn't being used for filtering.